src/metric.rs

```diff
@@ -0,0 +1,305 @@
+use byteorder::WriteBytesExt;
+use memmap::{Mmap, MmapViewSync, Protection};
+use std::ffi::CString;
+use std::io;
+use std::io::Cursor;
+use std::mem;
+
+const ITEM_BIT_LEN: usize = 10;
+
+pub const I32_METRIC_TYPE_CODE: u32 = 0;
+pub const U32_METRIC_TYPE_CODE: u32 = 1;
+pub const I64_METRIC_TYPE_CODE: u32 = 2;
+pub const U64_METRIC_TYPE_CODE: u32 = 3;
+pub const F32_METRIC_TYPE_CODE: u32 = 4;
+pub const F64_METRIC_TYPE_CODE: u32 = 5;
+pub const STRING_METRIC_TYPE_CODE: u32 = 6;
+
+pub trait MetricType {
+    fn type_code(&self) -> u32;
+    fn write_to_writer<W: WriteBytesExt>(&self, writer: &mut W) -> io::Result<()>;
+}
+
+macro_rules! impl_metric_type_for (
+    ($typ:tt, $base_typ:tt, $type_code:expr) => (
+        impl MetricType for $typ {
+            
+            fn type_code(&self) -> u32 {
+                $type_code
+            }
+
+            fn write_to_writer<W: WriteBytesExt>(&self, mut w: &mut W) -> io::Result<()> {
+                w.write_u64::<super::Endian>(
+                    unsafe {
+                        mem::transmute::<$typ, $base_typ>(*self) as u64
+                    }
+                )
+            }
+
+        }
+    )
+);
+
+impl_metric_type_for!(i32, u32, I32_METRIC_TYPE_CODE);
+impl_metric_type_for!(u32, u32, U32_METRIC_TYPE_CODE);
+impl_metric_type_for!(i64, u64, I64_METRIC_TYPE_CODE);
+impl_metric_type_for!(u64, u64, U64_METRIC_TYPE_CODE);
+impl_metric_type_for!(f32, u32, F32_METRIC_TYPE_CODE);
+impl_metric_type_for!(f64, u64, F64_METRIC_TYPE_CODE);
+
+impl MetricType for String {
+    fn type_code(&self) -> u32 {
+        STRING_METRIC_TYPE_CODE
+    }
+
+    fn write_to_writer<W: WriteBytesExt>(&self, mut writer: &mut W) -> io::Result<()> {
+        writer.write_all(CString::new(self.as_str())?.to_bytes_with_nul())
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum SpaceScale {
+    /// byte
+    Byte = 0,
+    /// kilobyte (1024) 
+    KByte,
+    /// megabyte (1024^2)
+    MByte,
+    /// gigabyte (1024^3)
+    GByte,
+    /// terabyte (1024^4)
+    TByte,
+    /// petabyte (1024^5)
+    PByte,
+    /// exabyte (1024^6)
+    EByte
+}
+
+#[derive(Copy, Clone)]
+pub enum TimeScale {
+    /// nanosecond
+    NSec = 0,
+    /// microsecond
+    USec,
+    /// millisecond
+    MSec,
+    /// second
+    Sec,
+    /// minute
+    Min,
+    /// hour
+    Hour
+}
+
+#[derive(Copy, Clone)]
+pub enum CountScale {
+    One = 0
+}
+
+
+pub struct Unit {
+    pub space_scale: SpaceScale,
+    pub time_scale: TimeScale,
+    pub count_scale: CountScale,
+    pub space_dim: i8,
+    pub time_dim: i8,
+    pub count_dim: i8
+}
```

Comment:
`Unit` should be a trait, so it can be extended freely. Again, Space, Time and Count Unit types can simply be type aliases, with individual variables holding runtime values.